libvast/vast/hash_index.hpp

```diff
@@ -0,0 +1,256 @@
+template <size_t Bytes>
+class hash_index : public value_index {
+  static_assert(Bytes > 0);
+  static_assert(Bytes <= 8);
+
+  /// The maximum number of hash rounds to try to find a new digest.
+  static constexpr size_t max_hash_rounds = 32;
+
+public:
+  using hasher_type = xxhash64;
+  using digest_type = std::array<byte, Bytes>;
+
+  static_assert(sizeof(hasher_type::result_type) >= sizeof(digest_type),
+                "number of chosen bytes exceeds underlying digest size");
+
+  /// Computes a chopped digest from arbitrary data.
+  /// @param x The data to hash.
+  /// @param seed The seed to use during the hash.
+  /// @returns The chopped digest.
+  static digest_type hash(data_view x, size_t seed = 0) {
+    digest_type result;
+    auto digest = uhash<hasher_type>{seed}(x);
+    std::memcpy(result.data(), &digest, Bytes);
+    return result;
+  }
+
+  /// Constructs a hash index for a particular type and digest cutoff.
+  /// @param t The type associated with this index.
+  /// @param digest_bytes The number of bytes to keep of a hash digest.
+  explicit hash_index(vast::type t) : value_index{std::move(t)} {
+  }
+
+  caf::error serialize(caf::serializer& sink) const override {
+    // Prune unneeded seeds.
+    decltype(seeds_) non_null_seeds;
+    for (auto& [k, v] : seeds_)
+      if (v > 0)
+        non_null_seeds.emplace(k, v);
+    return caf::error::eval([&] { return value_index::serialize(sink); },
+                            [&] { return sink(digests_, non_null_seeds); });
+  }
+
+  caf::error deserialize(caf::deserializer& source) override {
+    return caf::error::eval([&] { return value_index::deserialize(source); },
+                            [&] { return source(digests_, seeds_); });
+  }
+
+private:
+  struct key {
+    friend bool operator==(key x, digest_type y) {
+      return std::memcmp(x.bytes.data(), y.data(), y.size()) == 0;
+    }
+
+    friend bool operator!=(key x, digest_type y) {
+      return !(x == y);
+    }
+
+    friend bool operator==(digest_type x, key y) {
+      return y == x;
+    }
+
+    friend bool operator!=(digest_type x, key y) {
+      return !(x == y);
+    }
+
+    friend bool operator==(key x, key y) {
+      return x == y.bytes;
+    }
+
+    friend bool operator!=(key x, key y) {
+      return !(x == y);
+    }
+
+    digest_type bytes;
+  };
+
+  struct key_hasher {
+    auto operator()(const key& x) const {
+      auto result = uint64_t{0};
+      std::memcpy(&result, x.bytes.data(), x.bytes.size());
+      return result;
+    }
+  };
+
+  // TODO: remove this function as of C++20.
+  auto find_seed(data_view x) const {
+    auto pred = [&](const auto& xs) { return is_equal(xs.first, x); };
+    auto& xs = as_vector(seeds_);
+    return std::find_if(xs.begin(), xs.end(), pred);
+  }
+
+  // Retrieves the unique digest for a given input or generates a new one.
+  caf::optional<key> make_digest(data_view x) {
+    size_t i = 0;
+    do {
```

Comment:
Sometimes, a raw for loop might be the correct way to do things. 

```cpp
for (size_t i = 0; i < max_hash_rounds; ++i)
```

This seems to express the retry-n-times intent much clearer than the do-while loop.